Add color channel filter to editor texture previews

Zylann opened the pull request:
Implements and closes https://github.com/godotengine/godot-proposals/issues/11254

https://github.com/user-attachments/assets/2905fc1d-9d26-40f2-9bc2-5b9a4e58dcb9

Zylann committed:
Add color channel filter to editor texture previews

ckaiser commented:
Attempted to test it but couldn't get the buttons to do anything, plus Forward+ and Mobile render modes produced error spew that looked like this:
![Screenshot 2024-12-08 160707](https://github.com/user-attachments/assets/39267c6c-356d-4cd1-9dd5-77de8f14b1f9)

while the compatibility rendered failed differently:
![image](https://github.com/user-attachments/assets/ac9dd8ed-5c91-4467-82cf-2a4637baacef)

Neither of these errors appeared when testing a build with the same configuration without this patch.

Additionally, I think the contrast of the buttons could be better in light mode and in general - it's a bit hard to distinguish when its hovered and pressed, as seen here:

https://github.com/user-attachments/assets/60f41931-44c0-4191-8de4-1d62e31de179

---

Specs:
Windows 11
AMD Ryzen 5700X3D
Geforce RTX 3070
compiled with `scons p=windows target=editor optimize=speed use_llvm=yes lto=thin scu_build=yes production=yes`

Zylann commented:
There was line in shaders I forgot to remove. I pushed the fix.

>  I think the contrast of the buttons could be better in light mode and in general - it's a bit hard to distinguish when its hovered and pressed, as seen here:

I wish it looked better too. But I'm not sure how to theme them better, the default appearance of buttons was already quite bad, so I had to make them flat, remove focus which was in the way, and use a PanelContainer to get a background, which also had no default so I picked gray. Suggestions welcome.

Also TextureLayered and Texture3D previews don't use a container as parent, so I had to use anchors. However it's not behaving as expected.
In my test project, this is what I expect:
![image](https://github.com/user-attachments/assets/162c2cc1-38de-42a9-87ab-435aa34935a0)
So I did this in C++:
```cpp
	channel_selector->set_anchors_preset(Control::PRESET_BOTTOM_LEFT);
```
But it resulted in this:
![image](https://github.com/user-attachments/assets/dd0867c1-38e8-4dde-a4e2-b8958100632c)
I wonder what I'm missing here.

Calinou commented:
> I'm also considering to hide the selector if the image has only one channel.

Yes, I'd hide it in that situation.

ckaiser commented:
Works as expected now, after the changes!

![image](https://github.com/user-attachments/assets/ccabc430-9b9b-4430-b4d4-33a896095e1a)
I prefer the solid colors when things are selected, but there's some weird borders with the styling right now that make it look a bit messy around the edges.

![image](https://github.com/user-attachments/assets/134363b6-da80-4e43-9b02-7016e26c72be)
The unselected contrast is unchaged in light mode, so still not ideal.

Zylann commented:
Part of the problem is that there doesnt seem to be a default background to go well with the current theme, so I don't know how I should fix that.

Calinou commented:
Try using `some_button->set_theme_type_variation("FlatButton");` to use the same styling as the buttons in the 3D editor viewport (or `"FlatMenuButton"` but that one is meant for MenuButtons).

Zylann commented:
Here is how it looks with `FlatButton`: 

https://github.com/user-attachments/assets/e8c8aa48-75a5-4ee0-8353-c97aa8d6eddd

Not really better

arkology commented:
Will it look better with https://github.com/godotengine/godot/pull/100569?
Also maybe you could try to override some theme styles for these buttons - for example, outline or shadows.

Zylann commented:
After a bit of investigation, it appears `PanelContainer` does have a stylebox, but it is invisible. I have no idea why that is the case.
So to workaround that, I tried setting up the background of `Tree`, combined with @Calinou 's suggestion:

![image](https://github.com/user-attachments/assets/df782869-68fb-4674-b658-b082c0b0eae7)

It still doesn't look good at all. It's hard to see which buttons are pressed.
I would need the same background color as the one used in the toolbar, as this is like a mini toolbar.

> Will it look better with https://github.com/godotengine/godot/pull/100569?

I don't think this will affect things much.

> Also maybe you could try to override some theme styles for these buttons - for example, outline or shadows.

Been trying to do that for most of the work on that PR, but that missing background is quite an issue.

Zylann commented:
This time I tried with `TabContainer` `panel` and reduced its content margin:
![image](https://github.com/user-attachments/assets/d1287431-dccb-41d6-9150-630443b00591)
![image](https://github.com/user-attachments/assets/4f4f89ab-fe20-4662-b38b-b658b9a42912)

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/21a6bd8dcac41cd7751dd8de7dffe736c0b8965c), it works as expected. Looks great :slightly_smiling_face: 

Code looks good to me.

Black (OLED) | Light | Default (200% editor scale)
-|-|-
![image](https://github.com/user-attachments/assets/13353caa-f344-4ab5-b0c7-4b810a345780) | ![image](https://github.com/user-attachments/assets/0c1e9b7d-ea4a-47fd-ab81-078036943f3a) | ![image](https://github.com/user-attachments/assets/d7f5cac0-4e76-45af-a1e0-391dc12d185b)

Calinou commented on core/typedefs.h:
```diff
@@ -227,6 +227,12 @@ constexpr T get_num_bits(T x) {
 	return floor_log2(x);
 }
 
+// Tells if a positive integer is a power of 2
```
```suggestion
// Returns `true` if a positive integer is a power of 2, `false` otherwise.
```

Calinou commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,131 @@
+	Button *button = memnew(Button);
+	button->set_text(p_text);
+	button->set_toggle_mode(true);
+	button->set_pressed(true);
+	// button->set_flat(true);
```
Remove commented out code:

```suggestion
	button->set_pressed(true);
```

Calinou commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,131 @@
+	button->set_pressed(true);
+	// button->set_flat(true);
+
+	// Don't show focus, it stands out too much and remains visible which can be confusing
```
```suggestion
	// Don't show focus, it stands out too much and remains visible which can be confusing.
```

Calinou commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,131 @@
+	// Don't show focus, it stands out too much and remains visible which can be confusing
+	button->add_theme_style_override("focus", memnew(StyleBoxEmpty()));
+
+	// Make it look similar to toolbar buttons
```
```suggestion
	// Make it look similar to toolbar buttons.
```

Calinou commented on editor/plugins/texture_3d_editor_plugin.cpp:
```diff
@@ -44,8 +45,29 @@ constexpr const char *texture_3d_shader = R"(
 
+	uniform vec4 u_channel_factors = vec4(1.0);
+
+	vec4 filter_preview_colors(vec4 input_color, vec4 factors) {
+		// Filter RGB
+		vec4 output_color = input_color * vec4(factors.rgb, input_color.a);
+
+		// Remove transparency when alpha is not enabled
+		output_color.a = mix(1.0, output_color.a, factors.a);
+
+		// Switch to opaque grayscale when visualizing only one channel
+		float csum = factors.r + factors.g + factors.b + factors.a;
+		float single = clamp(2.0 - csum, 0.0, 1.0);
+		for (int i = 0; i < 4; ++i) {
+			float c = input_color[i];
+			output_color = mix(output_color, vec4(c, c, c, 1.0), factors[i] * single);
+		}
+
+		return output_color;
+	}
```
I think `i++` acts the same here, and we prefer using post-increment when it acts the same as preincrement.

```suggestion
	vec4 filter_preview_colors(vec4 input_color, vec4 factors) {
		// Filter RGB.
		vec4 output_color = input_color * vec4(factors.rgb, input_color.a);

		// Remove transparency when alpha is not enabled.
		output_color.a = mix(1.0, output_color.a, factors.a);

		// Switch to opaque grayscale when visualizing only one channel.
		float csum = factors.r + factors.g + factors.b + factors.a;
		float single = clamp(2.0 - csum, 0.0, 1.0);
		for (int i = 0; i < 4; i++) {
			float c = input_color[i];
			output_color = mix(output_color, vec4(c, c, c, 1.0), factors[i] * single);
		}

		return output_color;
	}
```

Calinou commented on editor/plugins/texture_3d_editor_plugin.cpp:
```diff
@@ -138,30 +162,44 @@ void Texture3DEditor::_update_gui() {
+
+	const uint32_t components_mask = Image::get_format_components_mask(format);
+	if (is_power_of_2(components_mask)) {
+		// Only one channel available, no point in showing a channel selector
```
```suggestion
		// Only one channel available, no point in showing a channel selector.
```

Calinou commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -39,6 +40,36 @@
+uniform vec4 u_channel_factors = vec4(1.0);
+
+vec4 filter_preview_colors(vec4 input_color, vec4 factors) {
+	// Filter RGB
+	vec4 output_color = input_color * vec4(factors.rgb, input_color.a);
+
+	// Remove transparency when alpha is not enabled
+	output_color.a = mix(1.0, output_color.a, factors.a);
```
```suggestion
	// Filter RGB.
	vec4 output_color = input_color * vec4(factors.rgb, input_color.a);

	// Remove transparency when alpha is not enabled.
	output_color.a = mix(1.0, output_color.a, factors.a);
```

Calinou commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -39,6 +40,36 @@
+	// Remove transparency when alpha is not enabled
+	output_color.a = mix(1.0, output_color.a, factors.a);
+
+	// Switch to opaque grayscale when visualizing only one channel
+	float csum = factors.r + factors.g + factors.b + factors.a;
+	float single = clamp(2.0 - csum, 0.0, 1.0);
+	for (int i = 0; i < 4; ++i) {
+		float c = input_color[i];
+		output_color = mix(output_color, vec4(c, c, c, 1.0), factors[i] * single);
+	}
```
```suggestion
	// Switch to opaque grayscale when visualizing only one channel.
	float csum = factors.r + factors.g + factors.b + factors.a;
	float single = clamp(2.0 - csum, 0.0, 1.0);
	for (int i = 0; i < 4; i++) {
		float c = input_color[i];
		output_color = mix(output_color, vec4(c, c, c, 1.0), factors[i] * single);
	}
```

Calinou commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -111,33 +168,62 @@ void TexturePreview::_update_metadata_label_text() {
+	texture_display->set_material(material);
 	add_child(texture_display);
 
+	// Null can be passed by `Camera3DPreview` (which immediately after sets a texture anyways)
```
```suggestion
	// Null can be passed by `Camera3DPreview` (which immediately after sets a texture anyways).
```

Calinou commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -111,33 +168,62 @@ void TexturePreview::_update_metadata_label_text() {
+	const Image::Format format = p_texture.is_valid() ? get_texture_2d_format(p_texture.ptr()) : Image::FORMAT_MAX;
+	const uint32_t components_mask = format != Image::FORMAT_MAX ? Image::get_format_components_mask(format) : 0xf;
+
+	// Add color channel selector at the bottom left if more than 1 channel is available
```
```suggestion
	// Add color channel selector at the bottom left if more than 1 channel is available.
```

arkology commented:
I'm worried that those buttons will always block part of texture preview.

Zylann commented:
@arkology I had the same worry, but it also applies to everything else we layer on top of those previews (details label, layer selector in 3D/layered textures).
If we really want to address that in this PR too, I thought these controls could be placed below the preview instead of being layered on top. I don't know how they should be laid out though.

arkology commented:
> but it also applies to everything else we layer on top of those previews

Oh yeah, something really needs to be done about this:
![Screenshot_2025-01-07-01-25-10-533_org godotengine editor v4 debug](https://github.com/user-attachments/assets/24cb1316-1afe-492c-b7b6-66a80dc4082c)

Calinou commented:
We can make the button background translucent (~50% opacity) so it's less distracting. The buttons should remain easily visible nonetheless.

arkology commented:
BTW please note that Camera3D preview and Subviewport will also be affected by this change.
And also not sure does it make sense to have the buttons in subresource preview like in comment above. I mean, how often it will be used?
I like this change, but it is hard to find proper place for it.

Zylann commented:
@Calinou this is how it looks like if I move it to the top and halve transparency:
![image](https://github.com/user-attachments/assets/a72dfacf-2508-47bc-aa9f-34b17deb4277)

@arkology Ah, forgot that TexturePreview has also been used for Camera3D and SubViewport... why is it the case? Barely anything is getting re-used, it just displays the texture and hides everything else. If the controls were not layered on top it could have been useful to have actually, but I guess I can hide them for now.
I'm not sure about removing that from the sub-inspectors. Those easily represent half of cases where I inspect stuff. And what about the other two controls? If we really want a safe spot, it could be below the preview. It will occupy a line though.

Zylann commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -66,25 +95,51 @@ void TexturePreview::_notification(int p_what) {
+	return Image::FORMAT_MAX;
+}
+
+static int get_texture_mipmaps_count(const Texture2D *texture) {
```
Will it elide extra calls to ref/unref?

AThousandShips replied:
Passing as `const Ref<Texture2D> &` should

Zylann commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -66,25 +95,51 @@ void TexturePreview::_notification(int p_what) {
-		format = Image::get_format_name(Object::cast_to<CompressedTexture2D>(*texture)->get_format());
-	} else {
-		format = texture->get_class();
+	const ImageTexture *image_texture = Object::cast_to<ImageTexture>(texture);
```
This also creates additional calls to ref/unref, but if using Ref is a blanket requirement then I guess I'll change that too

Zylann commented:
![image](https://github.com/user-attachments/assets/93eeded6-63ea-445c-9791-4626d3c1341b)
No idea what this means Oo My branch only has one commit:
![image](https://github.com/user-attachments/assets/ddf0d129-54f6-4dd8-bfef-ae046be1b3f1)

Calinou commented:
@Zylann Strange, this is what I see on my end at the bottom of the PR:

![image](https://github.com/user-attachments/assets/026d7eac-2370-4cf7-b091-3e419e11f8b2)

Zylann commented:
Ah, I no longer see it either. Not sure what happened.

akien-mga commented on core/typedefs.h:
```diff
@@ -222,6 +222,12 @@ constexpr T get_num_bits(T x) {
 	return floor_log2(x);
 }
 
+// Returns `true` if a positive integer is a power of 2, `false` otherwise.
+template <typename T>
+inline bool is_power_of_2(const T x) {
+	return x && ((x & (x - 1)) == 0);
+}
+
```
I'd move this further up next to the other `power_of_2` methods (probably as the first one).

This can be refactored to use `is_power_of_2`:
```
bool Image::is_size_po2() const {
        return uint32_t(width) == next_power_of_2(width) && uint32_t(height) == next_power_of_2(height);
}
```

And `core/os/memory.cpp` already had its own `is_power_of_2` which should be replaced by this one.

I also wonder whether this is going to raise warnings when used with unsigned types (the one in `core/os/memory.cpp` for `size_t` specifically uses unsigned literals:
```
inline bool is_power_of_2(size_t x) {
        return x && ((x & (x - 1U)) == 0U);
}
```

CC @hpvb

Zylann replied:
Would it make it more explicit to write it as `x && ((x & (x - T(1))) == T(0));`?

akien-mga commented on core/io/image.cpp:
```diff
@@ -3953,6 +3953,97 @@ String Image::get_format_name(Format p_format) {
+		case FORMAT_ASTC_8x8_HDR:
+			return rgba;
+		default:
+			ERR_PRINT("Unhandled format");
```
```suggestion
			ERR_PRINT("Unhandled format.");
```

akien-mga commented on core/io/image.cpp:
```diff
@@ -3953,6 +3953,97 @@ String Image::get_format_name(Format p_format) {
 	return format_names[p_format];
 }
 
+uint32_t Image::get_format_component_mask(Format p_format) {
+	const uint32_t r = 1;
+	const uint32_t rg = 3;
+	const uint32_t rgb = 7;
+	const uint32_t rgba = 15;
+
+	switch (p_format) {
+		case FORMAT_L8:
+			return rgb;
+		case FORMAT_LA8:
+			return rgba;
+		case FORMAT_R8:
+			return r;
+		case FORMAT_RG8:
+			return rg;
+		case FORMAT_RGB8:
+			return rgb;
+		case FORMAT_RGBA8:
+			return rgba;
```
CC @BlueCube3310, I have the feeling we might have logic for this kind of stuff already.

BlueCube3310 replied:
We don't have functions that would fit this exact purpose, so IMO this is fine to implement

akien-mga commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,134 @@
+#include "scene/resources/style_box_flat.h"
+
+ColorChannelSelector::ColorChannelSelector() :
+		buttons() {
```
Does that zero it properly? Usually we'd do `= {};` in the header directly.

Zylann replied:
I did that precisely for zero init, but `= {}` in header actually sounds better.

akien-mga commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,134 @@
+
+void ColorChannelSelector::_notification(int p_what) {
+	if (p_what == NOTIFICATION_THEME_CHANGED) {
+		// It is a common pattern in the editor to use `NOTIFICATION_THEME_CHANGED` to setup theme overrides that need
+		// access to the editor's theme. Unfortunately, in our case, this leads to infinite recursion.
```
What makes this case different from other editor classes?

CC @KoBeWi

AThousandShips commented on editor/plugins/color_channel_selector.h:
```diff
@@ -0,0 +1,59 @@
+public:
+	ColorChannelSelector();
+
+	void set_available_channels_mask(const uint32_t p_mask);
```
Not written anywhere explicitly but it's an established rule

akien-mga commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,134 @@
+}
+
+void ColorChannelSelector::create_button(unsigned int p_channel_index, const String &p_text, Control *p_parent) {
+	CRASH_COND(p_channel_index >= CHANNEL_COUNT);
+	CRASH_COND(buttons[p_channel_index] != nullptr);
```
`CRASH_COND` should be used extremely sparingly. Making the engine crash for end users because developers messed up is not the Godot philosophy.

These should be `ERR_FAIL_COND` most likely.

Zylann replied:
I used `CRASH_COND` here because this function is not supposed to even be called in any other way than in the constructor (it was for code re-use), and it's extremely obvious at development time if that would fail. There is no user code path that would cause it. I could as well remove it by now since I tested the code quite a few times, I only left it to keep the function obvious.
I can replace it to be totally safe, or wrap it in `DEV_ENABLED`.

akien-mga replied:
I'd say `ERR_FAIL_COND` should still be fine for this, especially if you don't foresee a risk that the contract would easily be broken. Alternatively `DEV_ASSERT`, but it still feels a bit heavy handed to me.

akien-mga commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -82,25 +113,51 @@ void TexturePreview::_update_texture_display_ratio() {
-void TexturePreview::_update_metadata_label_text() {
-	const Ref<Texture2D> texture = texture_display->get_texture();
+static Image::Format get_texture_2d_format(const Ref<Texture2D> &p_texture) {
+	// Why is format not just available in the Texture class?
```
Questions should be asked on chat or issues, not code :)

Zylann replied:
I think I implicitely wanted it to come up, so... here it is. Either we could add that right now, or could be done in another PR.

akien-mga commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -82,25 +113,51 @@ void TexturePreview::_update_texture_display_ratio() {
+		return compressed_texture->get_format();
+	}
+
+	// AtlasTexture?
```
Should be implemented.
CC @KoBeWi

Zylann replied:
I didn't actually implement it here because the original code treated AtlasTexture differently too. But it depends what we would prefer, maybe it could use the format of the atlas.

KoBeWi replied:
Using atlas' format makes sense.

akien-mga commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -82,25 +113,51 @@ void TexturePreview::_update_texture_display_ratio() {
+
+static int get_texture_mipmaps_count(const Ref<Texture2D> &p_texture) {
+	ERR_FAIL_COND_V(p_texture.is_null(), -1);
+	// We are having to download the image only to get its mipmaps count. It would be nice if we didn't have to.
```
CC @BlueCube3310

BlueCube3310 replied:
Since all images are (currently) expected to either have a complete mip chain or no mipmaps at all, this could be changed to `Image::get_image_required_mipmaps(...)` with the parameters from the `Texture2D` resource. In the future, we could implement the `get_mipmap_count` methods in each `Texture` class

BlueCube3310 replied:
Actually, nevermind. It's not even possible to check if a Texture2D has mipmaps, since unlike the layered and 3d variants it doesn't have a `has_mipmaps` vertual method. This should probably be added (not in this PR though)

KoBeWi replied:
btw "It would be nice" makes this eligible for TODO comment.

Zylann commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,134 @@
+
+void ColorChannelSelector::set_available_channels_mask(uint32_t p_mask) {
+	for (unsigned int i = 0; i < CHANNEL_COUNT; ++i) {
+		const bool available = (p_mask & (1 << i)) != 0;
```
I don't think I've encountered warnings for doing this, I used that quite a few times.

Zylann commented:
I'm agreeing a bit with the fact this control might be a little too present, even with half-transparency.
Do you think it would be better to put it behind a toggle like this?

![image](https://github.com/user-attachments/assets/c2e2497c-3647-430e-a219-be928a4e34f5)

arkology commented:
> Do you think it would be better to put it behind a toggle like this?

I have an idea that in future there could be context menu where visibility of such things as this one and metadata label (and so on) could be toggled. Like, you press RMB on texture and context menu appears.
But for now I think that for now hiding these buttons behind toggle is the solution.
Perhaps they don't even need to be created in constructor, but when you press a switch (maybe this is already done, I didn't look at code ATM).

BlueCube3310 commented:
> I'm agreeing a bit with the fact this control might be a little too present, even with half-transparency. Do you think it would be better to put it behind a toggle like this?

IMO it should be, the image previews are already pretty small, so space should be conserved as much as possible.

Zylann commented:
Changed to wrap with a toggle button. I added a new icon for it.
![image](https://github.com/user-attachments/assets/05d96f9b-bb57-4e1d-a611-b54eca3a14c0)

I also realized the outline added by #100569 is drawn with the texture preview itself, so the filter ends up applying to it too. So I made it drawn by another control on top.

arkology commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,146 @@
+	button->add_theme_style_override("focus", memnew(StyleBoxEmpty));
+
+	// Make it look similar to toolbar buttons.
+	button->set_theme_type_variation("FlatButton");
```
```suggestion
	button->set_theme_type_variation(SceneStringName(FlatButton));
```

arkology commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,146 @@
+	toggle_button = memnew(Button);
+	toggle_button->set_flat(true);
+	toggle_button->set_toggle_mode(true);
+	toggle_button->connect("toggled", callable_mp(this, &ColorChannelSelector::on_toggled));
```
```suggestion
	toggle_button->connect(SceneStringName(toggled), callable_mp(this, &ColorChannelSelector::on_toggled));
```

KoBeWi commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,146 @@
+	toggle_button->set_flat(true);
+	toggle_button->set_toggle_mode(true);
+	toggle_button->connect("toggled", callable_mp(this, &ColorChannelSelector::on_toggled));
+	toggle_button->add_theme_style_override("focus", memnew(StyleBoxEmpty));
```
Shouldn't this use FOCUS_MODE_NONE instead? Otherwise there is a button that can be focused without any visual feedback.

Zylann replied:
I checked similar buttons from `MaterialEditor`, and they don't use that, they also have a theme with empty focus stylebox.

KoBeWi replied:
:/

KoBeWi commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,146 @@
+	button->set_pressed(true);
+
+	// Don't show focus, it stands out too much and remains visible which can be confusing.
+	button->add_theme_style_override("focus", memnew(StyleBoxEmpty));
```
Same thing with focus.

KoBeWi commented on editor/plugins/texture_editor_plugin.cpp:
```diff
@@ -163,19 +222,48 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
 	centering_container->add_child(checkerboard);
 
+	{
+		Ref<Shader> shader;
+		shader.instantiate();
+		shader->set_code(texture_2d_shader);
```
It would be nice to not compile the shader every time. Though fixing it would require static shader like in ColorPicker, so it's something for the future.
Also now that we have instance uniforms, the material can be shared too.

Zylann replied:
Also applies to `Texture3DEditorPlugin` and `TextureLayeredEditorPlugin`

KoBeWi replied:
Yep.

arkology replied:
@KoBeWi I did it, as suggested, same way as `ColorPicker` - https://github.com/godotengine/godot/pull/104488

KoBeWi commented on editor/plugins/color_channel_selector.cpp:
```diff
@@ -0,0 +1,146 @@
+		ERR_FAIL_COND(bg_style.is_null());
+		bg_style = bg_style->duplicate();
+		// The default content margin makes the widget become a bit too large. It should be like mini-toolbar.
+		const float editor_scale = EditorScale::get_scale();
```
Normally we just use EDSCALE.

akien-mga commented:
Thanks!
